Rewrite description array.prototype.reduce

jenweber opened the pull request:
<!-- 👀 Thanks for opening a PR! Read comments like this one to get your PR merged faster. -->
#### Summary
<!-- ✍️ In a sentence or two, describe your changes -->

The quote from ECMA spec had a lot of redundant information that was covered in the Parameters section.
This removes redundant information and attempts to provide prose instead of pure specification.

#### Motivation
<!-- ❓ Why are you making this change? Help us understand how your changes help readers. -->

Reduce is very tricky for new developers. My aim is to make it easier to digest the information without sugar-coating or hiding the complexities. This follows my previous work here https://github.com/mdn/interactive-examples/pull/1996

#### Supporting details

Prior work by other contributors: https://github.com/mdn/content/pull/10555

Comment by @wbamberg  https://github.com/mdn/content/pull/10555: 
> I would love to replace that big quote from the spec

<!-- 🔗 Link to supporting information, such as bug trackers, source control, release notes, and vendor docs. -->

#### Related issues
Closes #12947 
<!-- 🔨 If applicable, use "Fixes #XYZ" -->

#### Metadata
<!-- ✅ Check a box if applicable, like this: [x]

This PR…
-->
- [ ] Adds a new document
- [x] Rewrites (or significantly expands) a document
- [ ] Fixes a typo, bug, or other error

<!-- 👷‍♀️ After submitting, review the results of the "Checks" tab! -->

### Changes

- Remove redundant information
- Categorize spec info about edge cases and mutations. 
- Give guidance about when to not use reduce.
- Organize examples - Moved 2 spec examples down to go with the rest of the examples.

### Open questions

Do we want to have the "when not to use" section? I saw something similar exists for [Array.prototype.map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map)

How's the ordering here?

jenweber committed:
Rewrite description array.prototype.reduce

- Remove redundant information
- Categorize spec info about edge cases andmutations. 
- Give guidance about when to not use reduce.
- Organize examples.

jenweber committed:
Slightly more details

sideshowbarker commented on files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:
```diff
@@ -77,24 +77,36 @@ The value that results from running the “reducer” callback function to compl
+`reduce` returns whatever is returned from the callback on the final iteration.
+
+### When to not use reduce()
+
+Recursive functions like `reduce` can be powerful, but sometimes difficult to understand,
+especially for less experienced JavaScript developers.
+If code becomes clearer when using other array methods,
+developers must weigh the readability tradeoff against the other benefits
+of using reduce.
+In cases where reduce is the best choice, documentation and semantic variable
+naming can help mitigate readability drawnbacks.
+
```
```suggestion

Recursive functions like `reduce` can be powerful, but sometimes difficult to understand,
especially for less-experienced JavaScript developers.
If code becomes more clear when using other array methods,
developers must weigh the readability tradeoff against the other benefits
of using `reduce()`.
In cases where `reduce()` is the best choice, documentation and well-chosen variable
names can help mitigate readability drawbacks.

```

dipikabh commented on files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:
```diff
@@ -77,24 +77,36 @@ The value that results from running the “reducer” callback function to compl
->
-> The range of elements processed by `reduce` is set before the first call to *callbackfn*. Elements that are appended to the array after the call to `reduce` begins will not be visited by *callbackfn*. If existing elements of the array are changed, their value as passed to *callbackfn* will be the value at the time `reduce` visits them; elements that are deleted after the call to `reduce` begins ...
+`reduce` takes two arguments, a callback function and an optional initial value.
+`reduce` calls the callback, as a function, once for each element in the array whenever
+an initial value is provided, in ascending order. If no initial value is provided,
+`reduce` calls the callback once for each element in the array after the first element.
```
```suggestion
If an initial value is provided, `reduce()` calls the "reducer" callback function on each element in the array, in order. If no initial value is provided, `reduce()` calls the callback function on each element in the array after the first element.
```
- Reworded the first sentence so that both sentences follow the same flow.
- Replaced "callback" with ""reducer" callback function" for consistency with the terminology used before this section
- The spec does say "ascending order". But as in the first paragraph on this page, saying "in order" will perhaps be less confusing (?).

dipikabh commented on files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:
```diff
@@ -77,24 +77,36 @@ The value that results from running the “reducer” callback function to compl
+
+### When to not use reduce()
+
+Recursive functions like `reduce` can be powerful, but sometimes difficult to understand,
+especially for less experienced JavaScript developers.
+If code becomes clearer when using other array methods,
+developers must weigh the readability tradeoff against the other benefits
+of using reduce.
+In cases where reduce is the best choice, documentation and semantic variable
+naming can help mitigate readability drawnbacks.
```
```suggestion
Recursive functions like `reduce()` can be powerful but sometimes difficult to understand,
especially for less experienced JavaScript developers.
If code becomes clearer when using other array methods,
developers must weigh the readability tradeoff against the other benefits
of using `reduce()`.
In cases where `reduce()` is the best choice, documentation and semantic variable
naming can help mitigate readability drawbacks.
```
- Removed the comma before "but" because here one of the two clauses being joined by "but" is not a complete sentence

dipikabh commented on files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:
```diff
@@ -77,24 +77,36 @@ The value that results from running the “reducer” callback function to compl
+`reduce` does not mutate the array it is used on, however it is possible for code inside the callback to mutate
+the array.
+
+If elements are appended to the array after `reduce` begins to iterate over the array, the appended
+elememts are not visited by *callbackfn*.
```
```suggestion
If elements are appended to the array after `reduce()` begins to iterate over the array, the appended
elements are not iterated by the callback function.
```
- Another rewrite suggestion to change the sentence to active voice:
"If elements are appended to the array after `reduce()` begins to iterate over the array, the callback function does not iterate over the appended elements."

jenweber replied:
I applied the second suggestion, thanks!

dipikabh commented on files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:
```diff
@@ -77,24 +77,36 @@ The value that results from running the “reducer” callback function to compl
+If elements are appended to the array after `reduce` begins to iterate over the array, the appended
+elememts are not visited by *callbackfn*.
+
+If existing elements of the array are changed, their value as passed to *callbackfn* will be the value at the time `reduce` visits them.
```
```suggestion
If existing elements of the array are changed, their values as passed to the callback function will be the values at the time `reduce()` iterates over the elements.
```

dipikabh commented on files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:
```diff
@@ -77,24 +77,36 @@ The value that results from running the “reducer” callback function to compl
+
+If existing elements of the array are changed, their value as passed to *callbackfn* will be the value at the time `reduce` visits them.
+
+Elements that are deleted after the call to `reduce` begins and before being visited are not visited.
```
```suggestion
Elements that are deleted after the call to `reduce()` begins and before being iterated over are not visited.
```

jenweber replied:
I think I want to take another pass in a followup PR to change some of this preexisting verbage. This is a pretty confusing sentence!

jenweber replied:
I made this a bulleted list now.

dipikabh reviewed (commented):
Hello @jenweber, thanks for this excellent writeup to replace the spec prose and also for categorising and highlighting the mutation info under its own subsection 👏🏻. Reorganising content to keep all examples together looks really good. IMO "When not to use" would be great info for developers! Thanks for adding this.

Apart from the edit suggestions (I have tried to explain the motivation behind the suggestions), I have replaced "reduce" with "reduce()". Elsewhere in the page, "reduce()" has been used, though I noticed that this convention has not been followed for [Array.propttype.map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map).

wbamberg reviewed (commented):
Thanks @jenweber for your PR and @dipikabh for the review, and sorry to be so slow getting to it.

Do you think it's worth trying to rescue/revise this bit of the original spec quote:

> callbackfn is called with four arguments:
>
> - the previousValue (value from the previous call to callbackfn)
> - the currentValue (value of the current element)
> - the currentIndex, and
> - the object being traversed
>
> The first time that callback is called, the previousValue and currentValue can be one of two values:
>
> - If an initialValue was supplied in the call to reduce, then previousValue will be equal to initialValue and currentValue will be equal to the first value in the array.
> - If no initialValue was supplied, then previousValue will be equal to the first value in the array and currentValue will be equal to the second.

? It seems useful, and everything else in the quote seems to be represented in your PR, as far as I could see.

I like the way you've used subsections to make the thing more scannable 👍 .

jenweber commented:
@wbamberg Thanks for the review. The reason I removed a big chunk of content is because the [Description](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce#description) and the [Parameters](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce#parameters) sections have duplicate content. I think that the Parameters section covers everything I removed.

Duplicate content:

<img width="824" alt="Screen Shot 2022-03-03 at 8 52 02 PM" src="https://user-images.githubusercontent.com/16627268/156683912-56d60201-ab6e-49af-85cc-081529f9008e.png">

<img width="806" alt="Screen Shot 2022-03-03 at 8 56 47 PM" src="https://user-images.githubusercontent.com/16627268/156684380-2144df83-fcc0-437e-8f46-a0a7c940e4b9.png">

jenweber committed:
Apply suggestions from code review

Thank you dipikabh!

Co-authored-by: Dipika Bhattacharya <dipika@foss-community.org>

jenweber committed:
Wording updates based on PR feedback

jenweber committed:
Use a bulleted list for mutations

jenweber commented:
@dipikabh thank you so much for all the readability edits. They are great!

I think I have applied & responded to all PR feedback, and this is ready for another review. Some possible followup steps after this PR would be to take one more pass at the readability. For the most part, I tried to reuse things as they were already written, but there is a lot of opportunity here to further improve the article.

dipikabh reviewed (changes requested):
Hi @jenweber, thanks for incorporating the suggestions! Yes, this definitely needed another pass for readability.

Let me know what you think about the suggestions. If you still feel that some sentence is not flowing naturally and would do better with a rewrite, do point it out and maybe we can work on it together.

For the array mutation section, I now feel it would be better to add examples to help readers visualize the three (complicated) scenarios. Apologies if I am increasing the scope of work.

wbamberg commented:
> @wbamberg Thanks for the review. The reason I removed a big chunk of content is because the [Description](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce#description) and the [Parameters](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce#parameters) sections have duplicate content. I think that the Parameters section covers everything I removed.
> 

Good point. I totally hadn't noticed that.

jenweber committed:
Apply suggestions from code review

Co-authored-by: Dipika Bhattacharya <dipika@foss-community.org>

jenweber commented on files/en-us/web/javascript/reference/global_objects/array/reduce/index.md:
```diff
@@ -77,24 +77,32 @@ The value that results from running the “reducer” callback function to compl
+
+- If elements are appended to the array after `reduce()` begins to iterate over the array, the callback
+function does not iterate over the appended elements.
+- If existing elements of the array are changed, their values as passed to the callback function will be the values at the time `reduce()` iterates over the elements.
```
@dipikabh let's have a look back at the spec and lean on that language here:

> If existing elements of the array are changed, their value as passed to callbackfn will be the value at the time reduce visits them

I see now that the prose above does not quite match the preciseness of this spec statement, mainly around the use of the word "iterate." Maybe we should put this back to be as close to the spec as we can, with teeny readability changes:

> If existing elements of the array do get changed, the values passed to the callback will be the values from the time that reduce was first called on the array.

How does that sound?

dipikabh replied:
That is definitely much better! Thanks for revisiting the spec and rewriting.
Just few minor updates:

> If existing elements of the array do get changed, the values passed to the callback function will be the values from the time that `reduce()` was first called on the array.

jenweber replied:
Updated!

jenweber commented:
Hi @dipikabh I would like to consider adding more examples and doing a full readability copy edit as out of scope for this PR. The goal of this PR is "remove redundant information and attempt to provide prose instead of pure specification." What do you think about doing that work in followup PRs? I'm interested in collaborating on a readability pass. I am collaborating with a junior dev as well who can provide UX feedback from that perspective.

With regards to giving examples for the edge cases - I am hesitant to give examples here because I think the best advice we could give devs is "don't modify the array as you are iterating over it" rather than helping them fully understand what happens if you do something weird. We could say something like that, but I don't have a great understanding of whether a section like this should be purely descriptive or also prescriptive. What do you think?

jenweber commented:
Also thanks for your feedback! I applied most of the suggestions and have 1 open question above for you, @dipikabh

dipikabh commented:
> Hi @dipikabh I would like to consider adding more examples and doing a full readability copy edit as out of scope for this PR. The goal of this PR is "remove redundant information and attempt to provide prose instead of pure specification." What do you think about doing that work in followup PRs?

💯 

> I'm interested in collaborating on a readability pass. I am collaborating with a junior dev as well who can provide UX feedback from that perspective.

Sure, let me know what you have in mind.

> With regards to giving examples for the edge cases - I am hesitant to give examples here because I think the best advice we could give devs is "don't modify the array as you are iterating over it" rather than helping them fully understand what happens if you do something weird. We could say something like that, but I don't have a great understanding of whether a section like this should be purely descriptive or also prescriptive. What do you think?

This is already a huge improvement. The mutation cases are edge cases, something you've already made devs aware of by documenting them. Since it was a bit complex for me to imagine an example when trying to read the mutation cases, I thought it might be nice to have the examples written. The examples in this particular case is just a nice-to-have feature, but definitely as I guessed out of scope of the current PR.

Thank you so much for your work on this PR! 🎉 
I guess only one tiny bit is pending and then we can merge this PR.

jenweber committed:
Apply PR feedback

jenweber commented:
I just applied the last update @dipikabh 

In terms of collaboration, I could open a copy edit PR and tag you. Or if you want to give it a try, I can review. I also pair program regularly with a junior dev who is interested to help out! We can ask them for feedback too. What do you think?

dipikabh commented:
> In terms of collaboration, I could open a copy edit PR and tag you. Or if you want to give it a try, I can review. I also pair program regularly with a junior dev who is interested to help out! We can ask them for feedback too. What do you think?

Sure, you can open a PR and tag me. This would be do to do another readthrough of this document, correct?

dipikabh reviewed (approved):
Thank you, this was great work! 👍

jenweber commented:
Yes, that’s right! I’ll open a ticket to describe the intent before I start on it, and tag you there too.

jenweber commented:
Thanks everyone!
